cmake/find/FindJulia.cmake

```
@@ -0,0 +1,108 @@
+macro(julia_bail_if_false message var)
+  if(NOT ${var})
+    set(Julia_FOUND 0)
+    set(Julia_NOT_FOUND_MESSAGE "${message}")
+    return()
+  endif()
+endmacro()
+
+# https://docs.julialang.org/en/v1/manual/environment-variables/#JULIA_BINDIR
+find_program(JULIA_EXECUTABLE julia PATHS ENV JULIA_BINDIR)
+julia_bail_if_false("The Julia executable could not be found" JULIA_EXECUTABLE)
+
+if(NOT DEFINED JULIA_BINDIR)
+  # The executable could be a chocolatey shim, so run some Julia code to report
+  # the path of the BINDIR
+  execute_process(
+      COMMAND "${JULIA_EXECUTABLE}" -e "print(Sys.BINDIR)"
+      OUTPUT_VARIABLE julia_bindir
+  )
+  file(TO_CMAKE_PATH "${julia_bindir}" julia_bindir)
+  set(JULIA_BINDIR "${julia_bindir}" CACHE PATH "")
+endif()
+julia_bail_if_false("The Julia executable could not report its location" JULIA_BINDIR)
+
+get_filename_component(julia_prefix "${JULIA_BINDIR}" DIRECTORY)
+
+if(WIN32)
+  set(julia_old_CMAKE_FIND_LIBRARY_SUFFIXES "")
+  set(julia_old_CMAKE_FIND_LIBRARY_PREFIXES "")
+  if(CMAKE_FIND_LIBRARY_SUFFIXES)
+    set(julia_old_CMAKE_FIND_LIBRARY_SUFFIXES "${CMAKE_FIND_LIBRARY_SUFFIXES}")
+    list(APPEND CMAKE_FIND_LIBRARY_SUFFIXES .dll.a)
+  else()
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".lib;.dll.a")
+  endif()
+  if(CMAKE_FIND_LIBRARY_PREFIXES)
+    set(julia_old_CMAKE_FIND_LIBRARY_PREFIXES "${CMAKE_FIND_LIBRARY_PREFIXES}")
+    list(APPEND CMAKE_FIND_LIBRARY_PREFIXES lib)
+  else()
+    set(CMAKE_FIND_LIBRARY_PREFIXES ";lib")
+  endif()
+endif()
```

Am I missing something or could we not just do
```suggestion
if (WIN32)
    set(CMAKE_FIND_LIBRARY_SUFFIXES "${CMAKE_FIND_LIBRARY_SUFFIXES}.lib;.dll.a")
endif()
```
?

---

Julia seems to be distributed with an MSYS2 build with library names that completely disregard Windows conventions. With normal Windows conventions, the library names should have been `julia.lib` for the import library and `julia.dll` for the DLL. Instead you get `libjulia.dll.a` and `libjulia.dll` respectively.

To fix this, we just have to make sure that `CMAKE_FIND_LIBRARY_PREFIXES` contains `lib` and `CMAKE_FIND_LIBRARY_SUFFIXES` contains `.dll.a`. However, if this ever gets fixed, I don't want this find module to break if it doesn't have to, so I make sure that it contains the values conventional to Windows and the ones currently in use for the Windows distribution of Julia.

The backup and restore here just makes sure that the people building jluna don't get their settings messed up and nothing else after calling this find module is affected.